Mathlib/LinearAlgebra/Matrix/Determinant/TotallyUnimodular.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2024 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak, Vladimir Kolmogorov, Ivan Sergeev
+-/
+import Mathlib.LinearAlgebra.Matrix.Determinant.Basic
+
+/-!
+# Totally unimodular matrices
+
+This file defines totally unimodular matrices and provides basic API for them.
+
+## Main definitions
+
+ - `Matrix.IsTotallyUnimodular`: a matrix is totally unimodular iff every square submatrix
+    (not necessarily contiguous) has determinant `0` or `1` or `-1`.
+
+## Main results
+
+ - `Matrix.isTotallyUnimodular_iff`: a matrix is totally unimodular iff every square submatrix
+    (possibly with repeated rows and/or repeated columns) has determinant `0` or `1` or `-1`.
+ - `Matrix.IsTotallyUnimodular.apply`: entry in a totally unimodular matrix is `0` or `1` or `-1`.
+
+-/
+
+open scoped Matrix
+
+variable {m n R : Type*} [CommRing R]
+
+/-- Is the matrix `A` totally unimodular? -/
+def Matrix.IsTotallyUnimodular (A : Matrix m n R) : Prop :=
+  ∀ k : ℕ, ∀ f : Fin k → m, ∀ g : Fin k → n, f.Injective → g.Injective →
+    (A.submatrix f g).det = 0 ∨
+    (A.submatrix f g).det = 1 ∨
+    (A.submatrix f g).det = -1
+
+lemma Matrix.isTotallyUnimodular_iff (A : Matrix m n R) : A.IsTotallyUnimodular ↔
+    ∀ k : ℕ, ∀ f : Fin k → m, ∀ g : Fin k → n,
+      (A.submatrix f g).det = 0 ∨
+      (A.submatrix f g).det = 1 ∨
+      (A.submatrix f g).det = -1 := by
+  constructor <;> intro hA
+  · intro k f g
+    if hf : f.Injective then
+      if hg : g.Injective then
+        exact hA k f g hf hg
+      else
+        left
+        unfold Function.Injective at hg
+        push_neg at hg
+        obtain ⟨i, j, hqij, hij⟩ := hg
+        apply Matrix.det_zero_of_column_eq hij
+        intro
+        simp [hqij]
+    else
+      left
+      unfold Function.Injective at hf
+      push_neg at hf
+      obtain ⟨i, j, hpij, hij⟩ := hf
+      apply Matrix.det_zero_of_row_eq
+      · exact hij
+      show (A (f i)) ∘ (g ·) = (A (f j)) ∘ (g ·)
+      rw [hpij]
```

Comment:
Can you use transpose matrices to deduplicate this?

Reply:
I'll try.

Reply:
I didn't find any easy way to do that.